File: examples/FinancialCalculator.js
/**
 * Calculadora Financeira
 * Implementa funções de cálculos financeiros comuns
 * @author Exemplo para Gerador de Testes Unitários
 */

class FinancialCalculator {
  /**
   * Calcula o valor futuro de um investimento
   * @param {number} principal - Valor inicial investido
   * @param {number} rate - Taxa de juros anual (em decimal, ex: 0.05 para 5%)
   * @param {number} time - Tempo em anos
   * @param {number} [contributions=0] - Contribuições periódicas anuais
   * @return {number} Valor futuro do investimento
   */
  calculateFutureValue(principal, rate, time, contributions = 0) {
    if (principal < 0 || rate < 0 || time < 0 || contributions < 0) {
      throw new Error("Todos os parâmetros devem ser valores não negativos");
    }
    
    // Cálculo do valor futuro com juros compostos
    const futureValuePrincipal = principal * Math.pow(1 + rate, time);
    
    // Cálculo das contribuições periódicas (assumindo final de período)
    const futureValueContributions = contributions * 
      ((Math.pow(1 + rate, time) - 1) / rate);
    
    return Number((futureValuePrincipal + futureValueContributions).toFixed(2));
  }
  
  /**
   * Calcula o valor da prestação de um empréstimo
   * @param {number} loanAmount - Valor total do empréstimo
   * @param {number} annualInterestRate - Taxa de juros anual (em decimal)
   * @param {number} loanTermYears - Prazo do empréstimo em anos
   * @return {number} Valor da prestação mensal
   */
  calculateLoanPayment(loanAmount, annualInterestRate, loanTermYears) {
    if (loanAmount <= 0 || annualInterestRate <= 0 || loanTermYears <= 0) {
      throw new Error("Os parâmetros devem ser valores positivos");
    }
    
    // Converter taxa de juros anual para mensal
    const monthlyRate = annualInterestRate / 12;
    
    // Converter prazo em anos para meses
    const termMonths = loanTermYears * 12;
    
    // Cálculo da prestação
    const payment = loanAmount * 
      (monthlyRate * Math.pow(1 + monthlyRate, termMonths)) / 
      (Math.pow(1 + monthlyRate, termMonths) - 1);
    
    return Number(payment.toFixed(2));
  }
  
  /**
   * Calcula o retorno sobre investimento (ROI)
   * @param {number} initialInvestment - Valor inicial investido
   * @param {number} finalValue - Valor final do investimento
   * @return {number} ROI em percentual
   */
  calculateROI(initialInvestment, finalValue) {
    if (initialInvestment <= 0) {
      throw new Error("O investimento inicial deve ser um valor positivo");
    }
    
    const roi = ((finalValue - initialInvestment) / initialInvestment) * 100;
    return Number(roi.toFixed(2));
  }
  
  /**
   * Calcula o valor presente de um montante futuro
   * @param {number} futureValue - Valor futuro
   * @param {number} rate - Taxa de juros anual (em decimal)
   * @param {number} time - Tempo em anos
   * @return {number} Valor presente
   */
  calculatePresentValue(futureValue, rate, time) {
    if (futureValue <= 0 || rate < 0 || time < 0) {
      throw new Error("Os parâmetros devem ser valores válidos");
    }
    
    const presentValue = futureValue / Math.pow(1 + rate, time);
    return Number(presentValue.toFixed(2));
  }
  
  /**
   * Calcula a taxa interna de retorno simples
   * @param {number} initialInvestment - Investimento inicial
   * @param {number} cashFlow - Fluxo de caixa anual constante
   * @param {number} years - Número de anos
   * @return {number} Taxa interna de retorno em percentual
   */
  calculateSimpleIRR(initialInvestment, cashFlow, years) {
    if (initialInvestment <= 0 || years <= 0) {
      throw new Error("O investimento inicial e anos devem ser valores positivos");
    }
    
    // Cálculo de IRR simples (aproximação)
    const totalCashFlow = cashFlow * years;
    const averageAnnualReturn = (totalCashFlow - initialInvestment) / years;
    const irr = (averageAnnualReturn / initialInvestment) * 100;
    
    return Number(irr.toFixed(2));
  }
}

/**
 * Função utilitária para formatar valores monetários
 * @param {number} value - Valor a ser formatado
 * @param {string} [currencyCode='BRL'] - Código da moeda
 * @param {string} [locale='pt-BR'] - Configuração regional
 * @return {string} Valor formatado como moeda
 */
function formatCurrency(value, currencyCode = 'BRL', locale = 'pt-BR') {
  return new Intl.NumberFormat(locale, {
    style: 'currency',
    currency: currencyCode
  }).format(value);
}

/**
 * Calcula a inflação acumulada
 * @param {number} initialValue - Valor inicial
 * @param {number} inflationRate - Taxa de inflação anual (em decimal)
 * @param {number} years - Número de anos
 * @return {number} Valor após inflação acumulada
 */
function calculateInflationImpact(initialValue, inflationRate, years) {
  if (initialValue < 0 || inflationRate < 0 || years < 0) {
    throw new Error("Os parâmetros devem ser valores não negativos");
  }
  
  const adjustedValue = initialValue * Math.pow(1 + inflationRate, years);
  return Number(adjustedValue.toFixed(2));
}

// Exportar para uso em outros módulos
module.exports = {
  FinancialCalculator,
  formatCurrency,
  calculateInflationImpact
};

File: examples/FinancialCalculator.java
/**
 * FinancialCalculator.java
 * Implementação de uma calculadora financeira com diversos métodos para cálculos financeiros comuns
 * 
 * @author Exemplo para Gerador de Testes Unitários
 * @version 1.0
 */

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.ArrayList;
import java.util.Currency;
import java.util.List;
import java.util.Locale;

public class FinancialCalculator {
    
    private static final int DECIMAL_PLACES = 2;
    private static final RoundingMode ROUNDING_MODE = RoundingMode.HALF_EVEN;
    
    /**
     * Calcula o valor futuro de um investimento com juros compostos
     * 
     * @param principal Valor inicial investido
     * @param rate Taxa de juros anual (em decimal, ex: 0.05 para 5%)
     * @param timeYears Tempo em anos
     * @param periodicContribution Contribuição periódica anual (opcional)
     * @return Valor futuro do investimento
     * @throws IllegalArgumentException se os parâmetros forem inválidos
     */
    public BigDecimal calculateFutureValue(BigDecimal principal, BigDecimal rate, 
                                          int timeYears, BigDecimal periodicContribution) {
        // Validação dos parâmetros
        if (principal == null || rate == null || periodicContribution == null) {
            throw new IllegalArgumentException("Os parâmetros não podem ser nulos");
        }
        
        if (principal.compareTo(BigDecimal.ZERO) < 0 || 
            rate.compareTo(BigDecimal.ZERO) < 0 || 
            timeYears < 0 ||
            periodicContribution.compareTo(BigDecimal.ZERO) < 0) {
            throw new IllegalArgumentException("Os parâmetros devem ser valores não negativos");
        }
        
        // Cálculo do valor futuro do principal
        BigDecimal onePlusRate = BigDecimal.ONE.add(rate);
        BigDecimal compoundFactor = onePlusRate.pow(timeYears);
        BigDecimal futureValuePrincipal = principal.multiply(compoundFactor);
        
        // Cálculo do valor futuro das contribuições periódicas
        BigDecimal futureValueContributions = BigDecimal.ZERO;
        if (periodicContribution.compareTo(BigDecimal.ZERO) > 0) {
            if (rate.compareTo(BigDecimal.ZERO) > 0) {
                BigDecimal rateFactorMinusOne = compoundFactor.subtract(BigDecimal.ONE);
                futureValueContributions = periodicContribution.multiply(rateFactorMinusOne.divide(rate, 10, ROUNDING_MODE));
            } else {
                // Se a taxa for zero, é apenas a multiplicação direta
                futureValueContributions = periodicContribution.multiply(new BigDecimal(timeYears));
            }
        }
        
        // Soma e arredondamento
        return futureValuePrincipal.add(futureValueContributions).setScale(DECIMAL_PLACES, ROUNDING_MODE);
    }
    
    /**
     * Sobrecarga do método calculateFutureValue sem contribuições periódicas
     */
    public BigDecimal calculateFutureValue(BigDecimal principal, BigDecimal rate, int timeYears) {
        return calculateFutureValue(principal, rate, timeYears, BigDecimal.ZERO);
    }
    
    /**
     * Calcula o valor da prestação de um empréstimo com base no sistema de amortização constante
     * 
     * @param loanAmount Valor total do empréstimo
     * @param annualInterestRate Taxa de juros anual (em decimal)
     * @param loanTermYears Prazo do empréstimo em anos
     * @return Valor da prestação mensal
     * @throws IllegalArgumentException se os parâmetros forem inválidos
     */
    public BigDecimal calculateLoanPayment(BigDecimal loanAmount, BigDecimal annualInterestRate, int loanTermYears) {
        // Validação dos parâmetros
        if (loanAmount == null || annualInterestRate == null) {
            throw new IllegalArgumentException("Os parâmetros não podem ser nulos");
        }
        
        if (loanAmount.compareTo(BigDecimal.ZERO) <= 0 || 
            annualInterestRate.compareTo(BigDecimal.ZERO) <= 0 || 
            loanTermYears <= 0) {
            throw new IllegalArgumentException("Os parâmetros devem ser valores positivos");
        }
        
        // Converter taxa de juros anual para mensal
        BigDecimal monthlyRate = annualInterestRate.divide(new BigDecimal("12"), 10, ROUNDING_MODE);
        
        // Converter prazo em anos para meses
        int termMonths = loanTermYears * 12;
        
        // Cálculo do fator de amortização
        BigDecimal onePlusRate = BigDecimal.ONE.add(monthlyRate);
        BigDecimal numerator = monthlyRate.multiply(onePlusRate.pow(termMonths));
        BigDecimal denominator = onePlusRate.pow(termMonths).subtract(BigDecimal.ONE);
        BigDecimal amortizationFactor = numerator.divide(denominator, 10, ROUNDING_MODE);
        
        // Cálculo da prestação
        return loanAmount.multiply(amortizationFactor).setScale(DECIMAL_PLACES, ROUNDING_MODE);
    }
    
    /**
     * Calcula o retorno sobre investimento (ROI)
     * 
     * @param initialInvestment Valor inicial investido
     * @param finalValue Valor final do investimento
     * @return ROI em percentual
     * @throws IllegalArgumentException se os parâmetros forem inválidos
     */
    public BigDecimal calculateROI(BigDecimal initialInvestment, BigDecimal finalValue) {
        // Validação dos parâmetros
        if (initialInvestment == null || finalValue == null) {
            throw new IllegalArgumentException("Os parâmetros não podem ser nulos");
        }
        
        if (initialInvestment.compareTo(BigDecimal.ZERO) <= 0) {
            throw new IllegalArgumentException("O investimento inicial deve ser um valor positivo");
        }
        
        // Cálculo do ROI em percentual
        BigDecimal profit = finalValue.subtract(initialInvestment);
        BigDecimal roi = profit.divide(initialInvestment, 10, ROUNDING_MODE).multiply(new BigDecimal("100"));
        
        return roi.setScale(DECIMAL_PLACES, ROUNDING_MODE);
    }
    
    /**
     * Calcula a taxa interna de retorno aproximada para fluxos de caixa uniformes
     * 
     * @param initialInvestment Investimento inicial (negativo)
     * @param cashFlows Lista de fluxos de caixa futuros
     * @return Taxa interna de retorno aproximada em percentual
     * @throws IllegalArgumentException se os parâmetros forem inválidos
     */
    public BigDecimal calculateApproximateIRR(BigDecimal initialInvestment, List<BigDecimal> cashFlows) {
        // Validação dos parâmetros
        if (initialInvestment == null || cashFlows == null || cashFlows.isEmpty()) {
            throw new IllegalArgumentException("Os parâmetros não podem ser nulos ou vazios");
        }
        
        if (initialInvestment.compareTo(BigDecimal.ZERO) >= 0) {
            throw new IllegalArgumentException("O investimento inicial deve ser um valor negativo");
        }
        
        // Soma de todos os fluxos de caixa
        BigDecimal totalCashFlow = BigDecimal.ZERO;
        for (BigDecimal cashFlow : cashFlows) {
            if (cashFlow == null) {
                throw new IllegalArgumentException("Os fluxos de caixa não podem conter valores nulos");
            }
            totalCashFlow = totalCashFlow.add(cashFlow);
        }
        
        // Calcula o valor médio dos fluxos de caixa
        BigDecimal averageCashFlow = totalCashFlow.divide(new BigDecimal(cashFlows.size()), 10, ROUNDING_MODE);
        
        // Cálculo aproximado da IRR (pressupõe fluxos de caixa uniformes)
        BigDecimal totalReturn = totalCashFlow.add(initialInvestment);
        BigDecimal yearsCount = new BigDecimal(cashFlows.size());
        BigDecimal annualReturn = totalReturn.divide(yearsCount, 10, ROUNDING_MODE);
        BigDecimal absoluteInitialInvestment = initialInvestment.abs();
        BigDecimal irr = annualReturn.divide(absoluteInitialInvestment, 10, ROUNDING_MODE).multiply(new BigDecimal("100"));
        
        return irr.setScale(DECIMAL_PLACES, ROUNDING_MODE);
    }
    
    /**
     * Formata um valor monetário de acordo com a localidade e moeda especificadas
     * 
     * @param amount Valor a ser formatado
     * @param locale Localidade a ser usada na formatação
     * @param currencyCode Código ISO 4217 da moeda
     * @return String formatada do valor monetário
     * @throws IllegalArgumentException se os parâmetros forem inválidos
     */
    public static String formatCurrency(BigDecimal amount, Locale locale, String currencyCode) {
        if (amount == null || locale == null || currencyCode == null || currencyCode.trim().isEmpty()) {
            throw new IllegalArgumentException("Os parâmetros não podem ser nulos ou vazios");
        }
        
        try {
            Currency currency = Currency.getInstance(currencyCode);
            java.text.NumberFormat formatter = java.text.NumberFormat.getCurrencyInstance(locale);
            formatter.setCurrency(currency);
            return formatter.format(amount);
        } catch (IllegalArgumentException e) {
            throw new IllegalArgumentException("Código de moeda inválido: " + currencyCode);
        }
    }
    
    /**
     * Calcula o valor presente de um montante futuro
     * 
     * @param futureValue Valor futuro
     * @param rate Taxa de juros anual (em decimal)
     * @param timeYears Tempo em anos
     * @return Valor presente
     * @throws IllegalArgumentException se os parâmetros forem inválidos
     */
    public BigDecimal calculatePresentValue(BigDecimal futureValue, BigDecimal rate, int timeYears) {
        // Validação dos parâmetros
        if (futureValue == null || rate == null) {
            throw new IllegalArgumentException("Os parâmetros não podem ser nulos");
        }
        
        if (futureValue.compareTo(BigDecimal.ZERO) <= 0 || 
            rate.compareTo(BigDecimal.ZERO) < 0 || 
            timeYears < 0) {
            throw new IllegalArgumentException("Os parâmetros devem ser valores válidos");
        }
        
        // Cálculo do valor presente
        BigDecimal onePlusRate = BigDecimal.ONE.add(rate);
        BigDecimal discountFactor = onePlusRate.pow(timeYears);
        BigDecimal presentValue = futureValue.divide(discountFactor, 10, ROUNDING_MODE);
        
        return presentValue.setScale(DECIMAL_PLACES, ROUNDING_MODE);
    }
}